sycl-fusion/passes/internalization/Internalization.cpp

```diff
@@ -437,112 +460,58 @@ void SYCLInternalizerImpl::promoteValue(Value *Val, std::size_t LocalSize,
   return Function::Create(NewTy, F->getLinkage(), F->getAddressSpace(),
                           F->getName(), F->getParent());
 }
 
-///
-/// Determine the element type of a pointer value by inspecting its uses. This
-/// is to determine the underlying type of a opaque pointer, whose element type
-/// is determined by its uses.
-static Type *getElementTypeFromUses(Value *PtrVal) {
-  assert(PtrVal->getType()->isPointerTy() && "Not a pointer type");
-  for (const auto &U : PtrVal->uses()) {
-    if (auto *I = dyn_cast<Instruction>(U.getUser())) {
-      Type *InferredTy = nullptr;
-      switch (I->getOpcode()) {
-      case Instruction::Call:
-      case Instruction::Invoke:
-      case Instruction::CallBr: {
-        auto *Call = cast<CallBase>(I);
-        auto Index = Call->getArgOperandNo(&U);
-        InferredTy =
-            getElementTypeFromUses(Call->getCalledFunction()->getArg(Index));
-        break;
-      }
-      case Instruction::AddrSpaceCast: {
-        InferredTy = getElementTypeFromUses(I);
-        break;
-      }
-      case Instruction::GetElementPtr:
-        InferredTy = cast<GetElementPtrInst>(I)->getSourceElementType();
-        break;
-      case Instruction::Load:
-        InferredTy = cast<LoadInst>(I)->getType();
-        break;
-      case Instruction::Store: {
-        auto *Store = cast<StoreInst>(I);
-        if (Store->getPointerOperand() == PtrVal) {
-          InferredTy = Store->getValueOperand()->getType();
-        }
-        break;
-      }
-      default:
-        // Unhandled case, rely on one of the following users to get type.
-        break;
-      }
-      if (InferredTy) {
-        return InferredTy;
-      }
-    }
-  }
-  return nullptr;
-}
-
 ///
 /// For private promotion, we want to replace each argument by an alloca.
-Value *replaceByNewAlloca(Argument *Arg, unsigned AS, std::size_t LocalSize) {
+Value *replaceByNewAlloca(Argument *Arg, unsigned AS, PromotionInfo PromInfo) {
   IRBuilder<> Builder{
       &*Arg->getParent()->getEntryBlock().getFirstInsertionPt()};
-  auto *PtrTy = cast<PointerType>(Arg->getType());
-  Type *Ty = getElementTypeFromUses(Arg);
-  assert(Ty && "Could not determine pointer element type");
-  auto *ArrTy = ArrayType::get(Ty, LocalSize);
-  auto *Alloca = Builder.CreateAlloca(ArrTy, PtrTy->getAddressSpace());
-  auto *Ptr = Builder.CreateInBoundsGEP(
-      ArrTy, Alloca, {Builder.getInt64(0), Builder.getInt64(0)});
-  Arg->replaceAllUsesWith(Ptr);
-  Alloca->mutateType(PointerType::getWithSamePointeeType(
-      cast<PointerType>(Alloca->getType()), AS));
-  Ptr->mutateType(PointerType::getWithSamePointeeType(
-      cast<PointerType>(Ptr->getType()), AS));
-  return Ptr;
+  auto ArgAS = cast<PointerType>(Arg->getType())->getAddressSpace();
+  auto *Alloca = Builder.CreateAlloca(
+      Builder.getInt8Ty(), ArgAS,
+      Builder.getInt64(PromInfo.LocalSize * PromInfo.ElemSize));
+  Alloca->setAlignment(Align{PromInfo.ElemSize});
```

Comment:
Does it make sense to set the alignment to the element size or should that be bounded somehow? The default would be 1, which is probably not what we want.

Reply:
But optimal alignment in global device memory may be different from optimal alignment in private or local memory. I'm not sure deriving it from the original pointer is the best idea here. Unless we can specifically get the required minimum alignment of the underlying type.

Reply:
I'd opt for using that value until we have a better option, but at least that might be better than 1. WDYT?

Reply:
Is the default really 1 here?

The LLVM LangRef says:
> If not specified, the target can choose to align the allocation on any convenient boundary compatible with the type.

I see two possible strategies here:
1. Use the original alignment, which would (presumably) match the alignment on any loads and stores using the original pointer. As those loads and stores will remain unchanged, this should work. 
2. Rely on the default.

As 2. has served us well so far, I would tend to say we should stick to 2.

Reply:
We discussed offline and chose (1).

Reply:
Should we set it to 1 if no alignment is given or just leave it unset?

Reply:
I think the result would be the same, because of the `i8` allocation, the default is `1` anyway.